From acb94c4a1ce4da027e9f7dbc370171148c7fc451 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tor=20Arne=20Vestb=C3=B8?= <tor.arne.vestbo@qt.io>
Date: Sun, 28 Sep 2025 23:39:28 +0200
Subject: [PATCH 372/553] QCocoaTheme: Resolve folder icons via NSWorkspace

The GetIconRef API we're using for the other theme icons is deprecated,
but sadly doesn't have a modern equivalent yet, as UTType doesn't provide
predefined UTTypes for all of the same icons, and we don't want to rely
on internal identifiers like com.apple.macbookpro-16-2023-space-gray in
place of kComputerIcon.

However for folder icons we can use NSWorkspace to either get an icon
for a specific folder, or the generic folder icon via UTTypeFolder.

This also automatically gives us the right tint for the folder icons
on macOS 26, reflecting the tint the user has chosen.

We no longer treat open and closed folders as visually different, as
macOS natively doesn't seem to do this anywhere I could find, and there
is no API to get the modern folder look in both these states with the
tint color preserved.

As before, we treat the "Link" variant of File and Folder icons as the
same as their non-link variants.

Pick-to: 6.8
Change-Id: If9c6ab4cffcf90b41b16adce338f494de4147629
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 381647427a5950f6bdeeb228068f4e3c92d656f5)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/plugins/platforms/cocoa/qcocoatheme.mm | 26 +++++++++++++++-------
 src/widgets/styles/qcommonstyle.cpp        |  2 ++
 2 files changed, 20 insertions(+), 8 deletions(-)

diff --git a/src/plugins/platforms/cocoa/qcocoatheme.mm b/src/plugins/platforms/cocoa/qcocoatheme.mm
index b784a7d362e..b1605d2c783 100644
--- a/src/plugins/platforms/cocoa/qcocoatheme.mm
+++ b/src/plugins/platforms/cocoa/qcocoatheme.mm
@@ -14,6 +14,7 @@
 #include "qcocoahelpers.h"
 
 #include <QtCore/qfileinfo.h>
+#include <QtCore/qstandardpaths.h>
 #include <QtCore/private/qcore_mac_p.h>
 #include <QtGui/private/qfont_p.h>
 #include <QtGui/private/qguiapplication_p.h>
@@ -34,6 +35,7 @@
 #include "qcocoamessagedialog.h"
 
 #include <CoreServices/CoreServices.h>
+#include <UniformTypeIdentifiers/UniformTypeIdentifiers.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -354,9 +356,16 @@ QPixmap QCocoaTheme::standardPixmap(StandardPixmap sp, const QSizeF &size) const
     case MessageBoxCritical:
         iconType = kAlertStopIcon;
         break;
-    case DesktopIcon:
-        iconType = kDesktopIcon;
-        break;
+    case DesktopIcon: {
+        auto desktop = QStandardPaths::writableLocation(QStandardPaths::DesktopLocation);
+        NSImage *icon = [NSWorkspace.sharedWorkspace iconForFile:desktop.toNSString()];
+        return qt_mac_toQPixmap(icon, size);
+    }
+    case DirHomeIcon: {
+        auto home = QStandardPaths::writableLocation(QStandardPaths::HomeLocation);
+        NSImage *icon = [NSWorkspace.sharedWorkspace iconForFile:home.toNSString()];
+        return qt_mac_toQPixmap(icon, size);
+    }
     case TrashIcon:
         iconType = kTrashIcon;
         break;
@@ -377,12 +386,13 @@ QPixmap QCocoaTheme::standardPixmap(StandardPixmap sp, const QSizeF &size) const
         iconType = kGenericNetworkIcon;
         break;
     case DirOpenIcon:
-        iconType = kOpenFolderIcon;
-        break;
+    case DirLinkOpenIcon:
+    case DirIcon:
     case DirClosedIcon:
-    case DirLinkIcon:
-        iconType = kGenericFolderIcon;
-        break;
+    case DirLinkIcon: {
+        NSImage *icon = [NSWorkspace.sharedWorkspace iconForContentType:UTTypeFolder];
+        return qt_mac_toQPixmap(icon, size);
+    }
     case FileLinkIcon:
     case FileIcon:
         iconType = kGenericDocumentIcon;
diff --git a/src/widgets/styles/qcommonstyle.cpp b/src/widgets/styles/qcommonstyle.cpp
index 129fd05d3d5..52f3eb1f0fa 100644
--- a/src/widgets/styles/qcommonstyle.cpp
+++ b/src/widgets/styles/qcommonstyle.cpp
@@ -5822,6 +5822,7 @@ QIcon QCommonStylePrivate::iconFromMacTheme(QCommonStyle::StandardPixmap standar
         case QStyle::SP_MessageBoxWarning:
         case QStyle::SP_MessageBoxCritical:
         case QStyle::SP_DesktopIcon:
+        case QStyle::SP_DirHomeIcon:
         case QStyle::SP_TrashIcon:
         case QStyle::SP_ComputerIcon:
         case QStyle::SP_DriveFDIcon:
@@ -5832,6 +5833,7 @@ QIcon QCommonStylePrivate::iconFromMacTheme(QCommonStyle::StandardPixmap standar
         case QStyle::SP_DirOpenIcon:
         case QStyle::SP_DirClosedIcon:
         case QStyle::SP_DirLinkIcon:
+        case QStyle::SP_DirLinkOpenIcon:
         case QStyle::SP_FileLinkIcon:
         case QStyle::SP_FileIcon:
         case QStyle::SP_ToolBarHorizontalExtensionButton:
-- 
2.51.2

